File: list-ops/ListOps.hs
{-# LANGUAGE BangPatterns #-}
module ListOps (
  (++),
  concat,
  filter,
  foldl',
  foldr,
  length,
  map,
  reverse
  ) where

import Control.Monad (ap)
import Data.Bool (bool)
import Prelude hiding ((++), concat, filter, foldr, length, map, reverse)

(++) :: [a] -> [a] -> [a]
xs ++ ys = foldr (:) ys xs

concat :: [[a]] -> [a]
concat = foldr (++) []

filter :: (a -> Bool) -> [a] -> [a]
filter p = foldr f []
  where f = ap (bool id . (:)) p

foldl' :: (b -> a -> b) -> b -> [a] -> b
foldl' f = lgo
  where lgo acc []      = acc
        lgo !acc (x:xs) = lgo (f acc x) xs

foldr :: (a -> b -> b) -> b -> [a] -> b
foldr f y = go
  where go [] = y
        go (x:xs) = f x $ go xs

-- | Illustration of the fold
--
-- @
--  length = go 0
--   where go acc []      = acc
--         go !acc (_:xs) = go (acc + 1) xs
-- @
--
length :: (Num b) => [a] -> b
length = foldl' (const . (1 +)) 0

map :: (a -> b) -> [a] -> [b]
map f = foldr g []
  where g =  (:) . f

reverse :: [a] -> [a]
reverse = foldl' (flip (:)) []

File: list-ops/list_ops.ml
let rec fold ~init:acc ~f = function
  | []        -> acc
  | (x :: xs) -> fold ~init:(f acc x) ~f xs

let length l =
  let rec go acc = function
    | [] -> acc
    | _ :: t -> go (acc+1) t in
  go 0 l

(* let length = fold ~init:0 ~f:(fun acc x -> acc + 1) *)

let reverse xs = fold ~init:[] ~f:(fun acc x -> x :: acc) xs

let map ~f xs =
  fold ~init:[] ~f:(fun acc x -> f x :: acc) xs
  |> reverse

let append xs ys =
  fold ~init:ys ~f:(fun acc x -> x :: acc) xs
  |> reverse

let concat xs =
  fold ~init:[] ~f:append xs |> reverse

let filter ~f xs =
  let g acc = function
    | x when f x -> x :: acc
    | _          -> acc in
  fold ~init:[] ~f:g xs |> reverse
